Source/WebCore/dom/InternalObserverFirst.cpp

```diff
@@ -0,0 +1,114 @@
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "InternalObserverFirst.h"
+
+#include "AbortController.h"
+#include "AbortSignal.h"
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortController>&& controller, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, WTFMove(controller), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = scriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref { m_promise }->resolve<IDLAny>(value);
+        Ref { m_controller }->abort(*JSC::jsCast<JSDOMGlobalObject*>(globalObject), JSC::jsUndefined());
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        Ref { m_promise }->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        Ref { m_promise }->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final
+    {
+    }
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortController>&& controller, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_controller(WTFMove(controller))
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Ref<AbortController> m_controller;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Observable& observable, const SubscribeOptions& options, Ref<DeferredPromise>&& promise)
+{
+    Ref controller = AbortController::create(context);
+
+    Vector<Ref<AbortSignal>> dependentSignals = { controller->protectedSignal() };
+    if (options.signal)
+        dependentSignals.append(*options.signal);
+    Ref dependentSignal = AbortSignal::any(context, dependentSignals);
+
+    if (dependentSignal->aborted())
+        return promise->reject<IDLAny>(dependentSignal->reason().getValue());
+
+    dependentSignal->addAlgorithm([promise](JSC::JSValue reason) {
+        promise->reject<IDLAny>(reason);
+    });
+
+    Ref observer = InternalObserverFirst::create(context, WTFMove(controller), WTFMove(promise));
+
+    observable.subscribeInternal(context, observer, SubscribeOptions { .signal = WTFMove(dependentSignal) });
```

Comment:
In terms of the failure, I haven't had time to debug yet but I suspect the issue is that nobody is keeping `dependentSignal` alive, so it gets destroyed before it gets aborted.

You're passing it to `observable.subscribeInternal()`, `subscribeInternal()` will add an algorithm to `dependentSignal` but it will not store it anywhere. As a result, I'd expect it to get destroyed.

Reply:
It appears I am right about the reason why the layout tests are failing. Nothing is keeping `dependentSignal` alive so it gets destroyed before we abort its child signal and thus it doesn't resolve the promise.

I have verified this is the cause by leaking the `dependentSignal`:
```
diff --git a/Source/WebCore/dom/InternalObserverFirst.cpp b/Source/WebCore/dom/InternalObserverFirst.cpp
index 1c83c3082a60..67b0e75b3314 100644
--- a/Source/WebCore/dom/InternalObserverFirst.cpp
+++ b/Source/WebCore/dom/InternalObserverFirst.cpp
@@ -98,6 +98,7 @@ void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Observ
     if (options.signal)
         dependentSignals.append(*options.signal);
     Ref dependentSignal = AbortSignal::any(context, dependentSignals);
+    dependentSignal->ref(); // LEAK!
 
     if (dependentSignal->aborted())
         return promise->reject<IDLAny>(dependentSignal->reason().getValue());
```

So someone needs to keep the dependentSignal alive. Hopefully this helps you come up with a proper fix.